### PR TITLE
🏗 Use Docker containers in CircleCI builds, and other optimizations

### PR DESCRIPTION
**Pull request by danielrozenberg**
This PR makes several optimizations to CircleCI builds that shave a "large PR" (i.e., one with the `RUNTIME` build target) from ~20 to ~16 minutes. Together with upcoming optimizations to Bundle Size, [Unit Tests](https://github.com/ampproject/amphtml/pull/35094), and [End-to-End Tests](https://github.com/ampproject/amphtml/pull/35158) we estimate this can go down to ~12 minutes eventually

This PR:
* Replaces VM executors with Docker containers, which [spin up immediately](https://circleci.com/docs/2.0/executor-types/#docker-benefits-and-limitations) (resolving the common complaint of "my PR is stuck on spinning up the environment for the past 20 minutes)
* Reuses CircleCI's built-in npm dependencies management steps, which are more efficient than simply calling `npm ci`
* Removes a dangling TODO from .sh files
* Splits Validator Tests to its own executor using a JRE-enabled Docker container (the Node container does not contain the JRE)
* Organizes steps in the .circleci/config.yml file
* Removes the 10 seconds delay for version checks in the `amp` command when running in a CI environment
* Disabled 13 e2e tests that became (even more) flaky (See #35241)

### CODE REVIEW

**Review by rileyajones — APPROVED** [review]

**Comment by danielrozenberg** [comment]
Note that this PR does cause E2E test to flake more often. I think it's safer to merge this and fix the flaky tests, instead of blocking an optimization due to flakes. wdyt? @rsimha @rileyajones

**Comment by rileyajones** [comment]
> Note that this PR does cause E2E test to flake more often. I think it's safer to merge this and fix the flaky tests, instead of blocking an optimization due to flakes. wdyt? @rsimha @rileyajones

@danielrozenberg do have any thoughts on the cause of the flakiness? It looks like the tests are flaking a majority of the time so I'm not sure it's worth it.

**Comment by danielrozenberg** [comment]
> @danielrozenberg do have any thoughts on the cause of the flakiness? It looks like the tests are flaking a majority of the time so I'm not sure it's worth it.

No idea, but it's probably to do with tests that are CPU intensive. It looks like it's the same ~dozen tests, so we can preemptively mark them for skip in this PR and assign them to be fixed

**Comment by danielrozenberg** [comment]
Aggregated list of e2e tests that failed on all the workflows of this PR:
```
amp carousel:0.2 in lightbox chrome Standalone environment should open with both arrows
amp carousel:0.2 in lightbox chrome Standalone environment should open with one arrow
amp-story-auto-ads:fullbleed chrome Standalone environment should render correctly
[v0.1] amp-base-carousel - snap property chrome Standalone environment should set snap property on spacers
[v1.0] amp-base-carousel - responsive attributes chrome Standalone environment should layout correctly after resize
[v1.0] amp-base-carousel - responsive attributes chrome Standalone environment should layout correctly initially
[v1.0] amp-base-carousel - responsive attributes chrome Standalone environment should respond to attribute changes
[v1.0] amp-base-carousel - responsive attributes chrome Standalone environment should retain position when changing the visible count
[v1.0] amp-social-share chrome Standalone environment rendering renders the default social share button
[v1.0] amp-social-share chrome Standalone environment rendering renders the social share button with custom type and endpoint
```

Come to think of it could this be a Chrome issue? This PR explicitly installs ChromeDriver, unless previous versions (this was required to make the `Checks` job run) so maybe that's related

File: .circleci/config.yml
Comment (rsimha) [inline .circleci/config.yml:78]: For maintainability, can we make this list opt-in instead of opt-out by invoking `install-chrome` like before and then calling `install-chromedriver`? 

(My concern is that if `install-browser-tools` adds a new default browser in future, we'll implicitly install it on every VM.)

File: .circleci/config.yml
Comment (rsimha) [inline .circleci/config.yml:66]: While there is definitely a speed advantage to using this orb, one of the visible downsides is that what used to be one step now occupies five steps, causing the job logs to be more noisy and less readable. This was why the original implementation uses a bash script that contains just what we need and nothing else.

**Before:**

image

**After:**

image

Curious: Would it be possible to do something similar here and retain the speed up while not growing the number of steps by extracting just the necessary steps from the orb?

**Review by rsimha — CHANGES_REQUESTED** [review]
Thanks for starting off this PR. Excited about the faster running times and the potential for $$ savings!

I've posted a few initial questions / suggestions below. I put in a blocking review because the task runner install change currently breaks local development.

File: build-system/task-runner/install-amp-task-runner.js
Comment (rsimha) [reply build-system/task-runner/install-amp-task-runner.js:37]: I did some more homework on this topic and remembered that we install the `amp` CLI runner to the global npm directory so that if a developer has multiple copies of the `amphtml` repo on their machine, the same global CLI runner copy (which is available on the system path once npm is installed) is used to invoke the repo-specific task.

If the `chown` workaround doesn't work elegantly, you could also try overriding the global npm prefix directory to a writable location on the docker container using `npm config set prefix path/to/writable/directory` before invoking `npm ci`.

Coupled with https://github.com/ampproject/amphtml/pull/35176#discussion_r668292374, I think there's good reason to reinstate `install_dependencies.sh` and distill an efficient / working set of commands.

Lemme know what you think. (This might be worth moving into its own PR.)

File: .circleci/config.yml
Comment (danielrozenberg) [reply .circleci/config.yml:66]: I understand the concern, but I think the payoff is worth the slight inconvenience. Honestly it would be great if CircleCI had an option to group steps together so that we can place all the setup steps together - maybe worth sending an FR?

File: .circleci/config.yml
Comment (danielrozenberg) [reply .circleci/config.yml:78]: browser-tools is version locked, so any additional browser will have to come from an explicit version upgrade. Is this still a concern in this case?

File: build-system/common/check-package-manager.js
Comment (danielrozenberg) [reply build-system/common/check-package-manager.js:286]: The `node` Docker instance doesn't have Python installed, and the `openjdk` Docker container has a patch-version number lag - in both cases, these aren't issues (should we even be testing for the presence of Python in any of the commands except for the validator? and, should we even be testing for the LTS version of Node in the case of the validator tests?)

File: build-system/task-runner/install-amp-task-runner.js
Comment (danielrozenberg) [reply build-system/task-runner/install-amp-task-runner.js:37]: Hmm... the thing is, it seems a bit "aggressive" to install anything to the global npm directory, it's not necessarily something that users would expect. Did you test the difference between --global and --user in a clean environment (with/without nvm?)

Comment (danielrozenberg) [reply build-system/task-runner/install-amp-task-runner.js]: In any case, this is something we can/should fix in a separate PR. I am restore the `--global` flag and added `npm config set prefix $HOME` which seems to fix the issue

File: .circleci/config.yml
Comment (danielrozenberg) [reply .circleci/config.yml]: I added commit 3d9b1fd258302223f5fffcc93a832affb22badf6, which merges several steps together to reduce overall steps count. wdyt?

File: .circleci/config.yml
Comment (rsimha) [reply .circleci/config.yml:73]: Version locking addresses the implicit-install concern, but doesn't address the maintainability concern. E.g. When we add new steps to be run only on the VMs that test on Firefox, etc. (coming soon once IE 11 is deprecated), opt-in steps will read better than opt-out.

**opt-out:**
```yaml
  install_chrome:
    steps:
      - browser-tools/install-browser-tools:
          install-chrome: true
          install-chromedriver: true
          install-firefox: false
          install-geckodriver: false
          replace-existing-chrome: true
  install_firefox:
    steps:
      - browser-tools/install-browser-tools:
          install-chrome: false
          install-chromedriver: false
          install-firefox: true
          install-geckodriver: false
          replace-existing-chrome: false
```

**opt-in:**
```yaml
  install_chrome:
    steps:
      - browser-tools/install-chrome:
          replace-existing: true
      - browser-tools/install-chromedriver
  install_firefox:
    steps:
      - browser-tools/install-firefox
```

File: build-system/common/check-package-manager.js
Comment (rsimha) [reply build-system/common/check-package-manager.js:286]: > The `node` Docker instance doesn't have Python installed

This used to be a problem once upon a time because our non-validator devtools required python. E.g. The nailgun closure compiler. Not a requirement these days.

> and the `openjdk` Docker container has a patch-version number lag

Validator tests directly invoke `python3`, so if that's available, we're good.

https://github.com/ampproject/amphtml/blob/0fde716d914c98d44ca41d93c1f2469b3a61c87e/build-system/tasks/validator.js#L30-L35
https://github.com/ampproject/amphtml/blob/0fde716d914c98d44ca41d93c1f2469b3a61c87e/build-system/tasks/validator.js#L64-L69

> should we even be testing for the LTS version of Node in the case of the validator tests?

> In any case, the delay doesn't really do anything in the CI

The `check-package-manager.js` script is meant to ensure that the environment is properly set up for local development. It has occasionally proved useful to debug issues during CI, so we still run it. But you've convinced me that the 10 second delay is useless during CI 😃 

My only suggestion: maybe simplify this block to return during CI without printing the "skipping delay prompt" message?

File: .circleci/config.yml
Comment (rsimha) [inline .circleci/config.yml:57]: I love the idea of consolidating these into one step. Perhaps shorten the name to something like "Setup Environment"?

File: .circleci/config.yml
Comment (rsimha) [inline .circleci/config.yml:48]: nit: Similar suggestion: Maybe shorten the step name to "Setup Workspace"?

File: build-system/task-runner/install-amp-task-runner.js
Comment (rsimha) [reply build-system/task-runner/install-amp-task-runner.js]: I did spend some time testing with the user npm bin dir when I first wrote the `amp` task runner. Nothing I tried would allow a developer to simply invoke `amp` without using workarounds like `npx amp` or `./amp.js`.

Maybe there's a way to achieve this by having `install-amp-task-runner.js` copy the `amp` runner to the user npm bin dir and then update the system path to point to it. (I couldn't figure out how to do this reliably in a platform independent way, but I'm guessing it's possible.)

Agree with trying to fix this separately 👍

File: .circleci/config.yml
Comment (rsimha) [reply .circleci/config.yml]: I like those other changes (thank you!), but still think we should attempt not to expand node package installation from 1 step to 5 steps. It's more troublesome to write this PR, but makes the job logs significantly easier to read, especially for new developers.

If speeding up package installation is independent from switching to docker containers, can we move this part into its own PR? WDYT?

File: build-system/common/check-package-manager.js
Comment (danielrozenberg) [reply build-system/common/check-package-manager.js:286]: I want to keep it exactly because it's the kind of helpful debugging statement that you just mentioned - if it shows up and we are experiencing deps issues, this would be a hint to look at

File: .circleci/config.yml
Comment (danielrozenberg) [reply .circleci/config.yml]: Would you mind if I submit this PR with this change and work on reducing the number of steps in a followup PR?

File: .circleci/config.yml
Comment (danielrozenberg) [reply .circleci/config.yml:77]: I'm convinced. Done.

File: .circleci/config.yml
Comment (danielrozenberg) [reply .circleci/config.yml:61]: We have too many steps with the title "setup" in them... I'm open to other suggestions :D

File: .circleci/config.yml
Comment (rsimha) [reply .circleci/config.yml]: No objections at all. Thanks for humoring me 😃

File: .circleci/config.yml
Comment (rsimha) [reply .circleci/config.yml:61]: Fair! How about "Configure environment"?

File: .circleci/config.yml
Comment (rsimha) [reply .circleci/config.yml:52]: And for this one, "Configure Workspace"?

File: build-system/common/check-package-manager.js
Comment (rsimha) [reply build-system/common/check-package-manager.js:286]: SGTM.

**Review by rsimha — APPROVED** [review]
LGTM now that everything else is just a nit 😃 

Thanks once again for speeding up CI!

File: .circleci/config.yml
Comment (danielrozenberg) [reply .circleci/config.yml:52]: `Configure Temporary Workspace`

File: .circleci/config.yml
Comment (danielrozenberg) [reply .circleci/config.yml:61]: Went with `Configure Development Environment`

File: .circleci/config.yml
Comment (rsimha) [inline .circleci/config.yml:32]: nit: Now that these two have moved from machine to docker, would it make sense to rename them to `amphtml-docker-{medium|large}-executor` to be consistent with `amphtml-docker-small-executor` above? Or alternatively, drop `docker-` from line 20?

File: .circleci/config.yml
Comment (rsimha) [inline .circleci/config.yml:37]: Similar nit here, re: consistently using / not using `docker-` in the executor name.

File: .circleci/config.yml
Comment (danielrozenberg) [reply .circleci/config.yml:32]: Yep, consistency is key. Let's remove the `-docker` tag, it doesn't really add much information given that the executor itself is defined with the keyword "docker" inside it. Done.

File: .circleci/config.yml
Comment (rsimha) [inline .circleci/config.yml:241]: I was going to suggest unblocking this PR by switching back to a linux machine just for the E2E tests and dealing with it separately. I then realized it's not only this job, but also the "Experiment A/B/C Tests" jobs that run E2E tests, so they'd all need to switch back to machine from docker.

May still be worth doing to unblock all the other improvements in this PR. No strong preference from me, so I'll leave this choice to you and @rileyajones.

Comment (danielrozenberg) [reply .circleci/config.yml:241]: I'm still trying things out, but I'm leaning towards keeping this `large` (there is no measurable improvement with `xlarge`) and disabling the dozen tests that became flaky. I'll re-run the workflow a few times to make sure this works as expected

Comment (danielrozenberg) [reply .circleci/config.yml]: btw, I tried switching to a machine, but it looks like there's some incompatibilities when moving the data between the Docker instances and the VM instances. Nothing that can't be resolved with enough hammers and nails, but I figured since we want to end up with insta-loading Docker containers anyways, I'd rather unblock this PR by skipping flakes rather than making a complicated startup script

Comment (rsimha) [reply .circleci/config.yml]: Good data point, and SGTM. 👍

Comment (rileyajones) [reply .circleci/config.yml]: This makes sense to me, the performance improvements are quite compelling.

A lot of the tests that are flaking with Docker appear to be commonly flaky tests anyway so skipping them might provide a good reason to look more closely at them individually later.

File: .circleci/config.yml
Comment (rsimha) [inline .circleci/config.yml:157]: Yet another thing I realized from looking at [this job](https://app.circleci.com/pipelines/github/ampproject/amphtml/12884/workflows/1453de09-26f5-44f9-bcfe-31b593e71fb5/jobs/206959): Today, the C++ validator tests are triggered only when validator code is changed. Since this PR isn't exercising that path, it might be worth triggering all tests once to make sure there are no surprises.

One way to do so is to push a temporary commit that touches any of our `package.json` files (doing so will trigger _all_ CI checks). Alternatively, merge this PR and observe the `main` branch push build, which will run all tests.

**Comment by rsimha** [comment]
Looping in @kristoferbaxter who requested that this be discussed during design review before we actually check in the change.

@danielrozenberg Would it be possible to write up an I2I for the move to docker containers and run it by `wg-approvers` to make sure there are no new considerations to think of?

**Comment by danielrozenberg** [comment]
> @danielrozenberg Would it be possible to write up an I2I for the move to docker containers and run it by `wg-approvers` to make sure there are no new considerations to think of? And if this meets the bar for a design review, can we use an upcoming time slot?

An I2I is a bit of an overkill - switching how we run our CI tasks is an implementation detail, there's no feature changes here. The discussion on flakes is here above and I think that's the only thing that seriously needs discussing :)

File: .circleci/config.yml
Comment (danielrozenberg) [reply .circleci/config.yml:157]: I tested running all targets and it passes: https://app.circleci.com/pipelines/github/ampproject/amphtml/13010/workflows/8b70cd3e-d6e3-466a-a422-4d9752a7c038

**Review by alanorozco — APPROVED** [review]
Approved skipping tests.

**Pull request merged**
